src/circuit/gadget/utilities/decompose_running_sum.rs

```diff
@@ -108,78 +98,101 @@ impl<
     /// Decompose a field element alpha that is witnessed in this helper.
     ///
     /// `strict` = true constrains the final running sum to be zero, i.e.
-    /// constrains alpha to be within WINDOW_NUM_BITS * NUM_WINDOWS bits.
+    /// constrains alpha to be within WINDOW_NUM_BITS * num_windows bits.
     pub fn witness_decompose(
         &self,
         region: &mut Region<'_, F>,
         offset: usize,
         alpha: Option<F>,
         strict: bool,
+        word_num_bits: usize,
+        num_windows: usize,
     ) -> Result<(CellValue<F>, RunningSum<F>), Error> {
         let z_0 = {
             let cell = region.assign_advice(
-                || "Witness alpha",
+                || "z_0 = alpha",
                 self.z,
                 offset,
                 || alpha.ok_or(Error::SynthesisError),
             )?;
             CellValue::new(cell, alpha)
         };
-        self.decompose(region, offset, z_0, strict)
+        self.decompose(region, offset, z_0, strict, word_num_bits, num_windows)
     }
 
     /// Decompose an existing variable alpha that is copied into this helper.
     ///
     /// `strict` = true constrains the final running sum to be zero, i.e.
-    /// constrains alpha to be within WINDOW_NUM_BITS * NUM_WINDOWS bits.
+    /// constrains alpha to be within WINDOW_NUM_BITS * num_windows bits.
     pub fn copy_decompose(
         &self,
         region: &mut Region<'_, F>,
         offset: usize,
         alpha: CellValue<F>,
         strict: bool,
+        word_num_bits: usize,
+        num_windows: usize,
     ) -> Result<(CellValue<F>, RunningSum<F>), Error> {
-        let z_0 = copy(region, || "Copy alpha", self.z, offset, &alpha, &self.perm)?;
-        self.decompose(region, offset, z_0, strict)
+        let z_0 = copy(region, || "z_0 = alpha", self.z, offset, &alpha, &self.perm)?;
+        self.decompose(region, offset, z_0, strict, word_num_bits, num_windows)
     }
 
+    /// `z_0` must be the cell at `(self.z, offset)` in `region`.
```

Comment:
If `z_0` must be this, why not query it rather than passing a `CellValue`?

Reply:
This is during assignment, not configuration. Here we need to know:
- it's value (if during proving).
- that it is assigned at the correct relative offset in the correct column (the precondition stated in the comment).

This method states a precondition because there are two wrapping APIs: one that witnesses the value directly, and one that copies it in from somewhere else.

Reply:
We could move the logic involving `z_0` out into the callers `witness_decompose()` / `copy_decompose()`, but I chose to include it in the internal function `decompose()` to reduce code duplication.

Reply:
I think it's fine as-is, now that the inner `decompose()` method is documenting the necessary precondition.

Reply:
Also, we can't _avoid_ that precondition, as the running sum intentionally uses relative offsets; even if we moved the `z_0` logic outside `decompose()`, the `z_1` logic would still require this precondition even if all we passed in was the value of `z_0`.